### PR TITLE
Pyrpp

### PR DESCRIPTION
**Pull request by HazarathKumarM**

### CODE REVIEW

File: rpp_pybind/README.md [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+pip install -e .
+```
+
+## Verify installation
+
+### Quick Verification
+
+```python
+import rpp_pybind as rpp
+from rpp_pybind import types
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply rpp_pybind/README.md]: Done.

File: rpp_pybind/README.md [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+
+# Check version
+print(f"PyRPP Version: {rpp.__version__}")
+
+# Check backend availability
+print(f"GPU Available: {types.is_gpu_available()}")
+print(f"Default Backend: {types.get_default_backend()}")
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply rpp_pybind/README.md]: Done

File: rpp_pybind/amd/rpp/rpp_types.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+HOST = RppBackend.HOST
+HIP = RppBackend.HIP
+
+NCHW = RpptLayout.NCHW
+NHWC = RpptLayout.NHWC
+
+U8 = RpptDataType.U8
+F32 = RpptDataType.F32
+F16 = RpptDataType.F16
+
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply rpp_pybind/amd/rpp/rpp_types.py:49]: Done.

File: CMakeLists.txt [hunk 4]
```diff
@@ -93,6 +93,17 @@ set(DEFAULT_BUILD_TYPE "Release")
...
 option(RPP_AUDIO_SUPPORT      "Build RPP with Audio Support"         ON)
 
+option(RPP_PYPACKAGE  "Build RPP Python package"  OFF)
+set(PYTHON_VERSION_SUGGESTED "" CACHE STRING "User provided Python version for RPP pybind")
+
+# Validate Python version if provided once
+if(PYTHON_VERSION_SUGGESTED)
+    if(PYTHON_VERSION_SUGGESTED AND (${PYTHON_VERSION_SUGGESTED} VERSION_LESS "3.8" OR ${PYTHON_VERSION_SUGGESTED} VERSION_GREATER "3.13"))
+        message("-- ${Red}NOTE: RPP pybind supported on Python versions - 3.8, 3.9, 3.10, 3.11, 3.12, 3.13. Please provide supported version${ColourReset}")
+        set(BUILD_PYPACKAGE OFF)
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply CMakeLists.txt]: Done

File: rpp_pybind/README.md [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+- `--num_runs`: Number of iterations for performance testing
+- `--case_list`: Specific augmentation cases to test
+
+## Usage Examples
+
+### Basic Example
+
+```python
+import rpp_pybind as rpp
+from rpp_pybind import fn, utils, types
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply rpp_pybind/README.md]: Done.

File: utilities/python_tests/test_suite.py [hunk 6]
```diff
@@ -0,0 +1,1475 @@
...
+            }
+
+            if self.test_type == 0 and self.qa_mode and self.qa_file:
+                self._write_aug_summary(aug_name, variant_name, timing_info)
+
+            print("\n" + "-" * 90)
+
+            # Record per-variant results
+            if self.test_type == 0 and self.qa_mode:
+                success = overall_success_count == overall_total
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply utilities/python_tests/test_suite.py]: Done.

File: rpp_pybind/rpp_pybind.cpp [hunk 7]
```diff
@@ -0,0 +1,941 @@
...
+        beta_ptr = beta_gpu_ptr;
+        roi_ptr = roi_gpu_ptr;
+    }
+    
+    rppt_brightness(input_data.ptr, &src_desc,
+                   output_data.ptr, &dst_desc,
+                   alpha_ptr,
+                   beta_ptr,
+                   roi_ptr, RpptRoiType::XYWH,
+                   rpp_handle, static_cast<RppBackend>(backend));
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply rpp_pybind/rpp_pybind.cpp]: Done

File: rpp_pybind/rpp_pybind.cpp [hunk 8]
```diff
@@ -0,0 +1,941 @@
...
+    for(int i = 0; i < tensor.dim(); i++) {
+        data.shape.push_back(tensor.size(i));
+        data.strides.push_back(tensor.stride(i));
+    }
+    
+    // Set device
+    if(tensor.is_cuda()) {
+        data.device = {kDLROCM, 0};  // ROCm device
+    } else {
+        data.device = {kDLCPU, 0};
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply rpp_pybind/rpp_pybind.cpp]: Done.

File: CMakeLists.txt [hunk 9]
```diff
@@ -422,6 +435,13 @@ install(EXPORT ${PROJECT_NAME}-targets
     COMPONENT dev
 )
 
+# Python Package
+if(RPP_PYPACKAGE)
+    add_subdirectory(rpp_pybind)
+else()
+    message("--${Yellow} RPP Python Module turned OFF by user option -D BUILD_PYPACKAGE=OFF or unsupported Python version${ColourReset}")
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply CMakeLists.txt]: Done.

File: rpp_pybind/amd/rpp/utils.py [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+
+
+def get_layout_name(layout_enum):
+    """Get string name for layout enum"""
+    from rpp_pybind.amd.rpp.rpp_types import RpptLayout
+    
+    if layout_enum == RpptLayout.NCHW:
+        return "PKD3"
+    elif layout_enum == RpptLayout.NHWC:
+        return "PLN3"
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply rpp_pybind/amd/rpp/utils.py]: Done.

File: rpp_pybind/amd/rpp/fn.py [hunk 11]
```diff
@@ -0,0 +1,572 @@
...
+    
+    batch_size = images.shape[0]
+    output = torch.empty_like(images)
+    
+    # Set ROI dimensions
+    if roi_widths is None:
+        roi_widths = [images.shape[3] if input_layout == 'NCHW' else images.shape[2]] * batch_size
+    if roi_heights is None:
+        roi_heights = [images.shape[2] if input_layout == 'NCHW' else images.shape[1]] * batch_size
+
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply rpp_pybind/amd/rpp/fn.py]: Done

File: rpp_pybind/amd/rpp/fn.py [hunk 12]
```diff
@@ -0,0 +1,572 @@
...
+    
+    # Move to correct device
+    if backend == HIP and not images.is_cuda:
+        images = images.cuda()
+    elif backend == HOST and images.is_cuda:
+        images = images.cpu()
+
+    batch_size = images.shape[0]
+    output = torch.zeros_like(images).contiguous()
+
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply rpp_pybind/amd/rpp/fn.py]: Done

File: rpp_pybind/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+    message("-- ${Green}RPP pybind module ${mod_tgt} added${ColourReset}")
+endforeach()
+
+# Install Python package files (pure Python)
+# NOTE: We must preserve the directory structure (amd/..., amd/rpp/...) so that
+# imports like `rpp_pybind.amd.rpp.fn` work.
+install(
+    FILES
+        "${CMAKE_CURRENT_SOURCE_DIR}/__init__.py"
+        "${CMAKE_CURRENT_SOURCE_DIR}/fn.py"
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply rpp_pybind/CMakeLists.txt:149]: Done.

File: rpp_pybind/rpp_pybind.cpp [hunk 14]
```diff
@@ -0,0 +1,980 @@
...
+
+    // Handle management
+    // reinterpret_cast<rppHandle_t>(static_cast<uintptr_t>(handle))
+    m.def("rppCreate", [](int batch_size, int backend) {
+        rppHandle_t handle = nullptr;
+        rppCreate(&handle,
+                  static_cast<size_t>(batch_size),
+                  0,
+                  nullptr,
+                  static_cast<RppBackend>(backend));
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply rpp_pybind/rpp_pybind.cpp]: Done.

File: rpp_pybind/rpp_pybind.cpp [hunk 15]
```diff
@@ -0,0 +1,980 @@
...
+                        alpha_ptr,
+                        beta_ptr,
+                        roi_ptr, RpptRoiType::XYWH,
+                        rpp_handle, static_cast<RppBackend>(backend));
+
+    if (status != RPP_SUCCESS) {
+        throw std::runtime_error("rppt_brightness failed");
+    }
+    
+    if(backend == 1)
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply rpp_pybind/rpp_pybind.cpp]: Done.

File: rpp_pybind/README.md [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+
+```python
+# Auto-detect backend (GPU if available, else CPU)
+output = fn.brightness(images, alpha=1.5)
+
+# Explicitly use CPU backend
+output = fn.brightness(images, alpha=1.5, backend=types.HOST)
+
+# Explicitly use GPU backend
+output = fn.brightness(images, alpha=1.5, backend=types.HIP)
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply rpp_pybind/README.md]: Done.

File: rpp_pybind/amd/rpp/fn.py [hunk 17]
```diff
@@ -0,0 +1,572 @@
...
+    
+    # Ensure tensor is contiguous and on correct device  
+    if not images.is_contiguous():
+        images = images.contiguous()
+    
+    # Move to correct device
+    if backend == HIP and not images.is_cuda:
+        images = images.cuda()
+    elif backend == HOST and images.is_cuda:
+        images = images.cpu()
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply rpp_pybind/amd/rpp/fn.py]: Done.

File: rpp_pybind/README.md [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+
+### Module Structure
+
+```
+rpp_pybind/
+├── __init__.py       # Main module with C++ exports
+├── fn.py            # High-level augmentation functions
+├── rpp_types.py     # Type definitions and helpers
+├── utils.py         # Image loading/saving utilities
+├── rpp_pybind.cpp   # C++ pybind11 bindings
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply rpp_pybind/README.md]: Done

File: utilities/python_tests/test_suite.py [hunk 19]
```diff
@@ -0,0 +1,1473 @@
...
+                                else:
+                                    output_for_qa = output
+                            else:
+                                output_for_qa = output
+
+                        passed, stats = self._compare_output(
+                            output_for_qa, ref_data, idx,
+                            is_grayscale=grayscale, aug_name=aug_name)
+
+                        if passed:
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply utilities/python_tests/test_suite.py:968]: Done.

File: rpp_pybind/amd/rpp/fn.py [hunk 20]
```diff
@@ -0,0 +1,572 @@
...
+
+    batch_size = images.shape[0]
+    output = torch.zeros_like(images).contiguous()
+    
+    # Set ROI dimensions
+    if roi_widths is None:
+        roi_widths = [images.shape[3] if images.shape[1] <= 3 else images.shape[2]] * batch_size
+    if roi_heights is None:
+        roi_heights = [images.shape[2] if images.shape[1] <= 3 else images.shape[1]] * batch_size
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply rpp_pybind/amd/rpp/fn.py]: Done.

File: utilities/python_tests/test_suite.py [hunk 21]
```diff
@@ -0,0 +1,1473 @@
...
+        return test_cases
+
+    def run_performance_tests(self):
+        """Run performance tests (test_type=1)."""
+        device            = 'cuda' if self.backend == rpp_type.HIP else 'cpu'
+        num_iterations    = self.num_runs if self.num_runs > 1 else 100
+        warmup_iterations = 5
+
+        base_image = load_image_with_bitdepth(self.test_images[1], bitdepth=self.bitdepth, device=device)
+        print(f"Running {num_iterations} iterations per augmentation "
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply utilities/python_tests/test_suite.py]: Done.

File: utilities/python_tests/test_suite.py [hunk 22]
```diff
@@ -0,0 +1,1473 @@
...
+                if qa_file:
+                    test_suite.set_shared_qa_file(qa_file, qa_output_dir)
+
+                if args.test_type == 1 and perf_log_file:
+                    test_suite.set_shared_perf_log(perf_log_file, perf_output_dir)
+                    perf_log_file.write(f"\n{'='*50}\nBitDepth: {bitdepth}\n{'='*50}\n")
+                    perf_log_file.flush()
+
+                success = test_suite.run_all()
+
```
[the remark this answers is not in the document]
Comment (mukeshj0606) [reply utilities/python_tests/test_suite.py:1441]: Done.
